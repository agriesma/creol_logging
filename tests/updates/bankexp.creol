/*
 * bankexp.creol
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2009 by Ingrid Chieh Yu de Vibe
 *   <ingridcy@ifi.uio.no>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Account
begin
  with Any
   // op increaseBalance(in s:Int)
    op balance(out b: Int)
    op deposit(in sum: Int)
end


class BankAccount implements Account 
 begin

   var bal: Int := 0;
   op increaseBalance (in sum : Int) == bal := bal + sum

   with Any
   op balance (out b : Int)==  b := bal
   op deposit(in sum : Int) == !increaseBalance(sum)
 end

class Customer
 begin
   var timer : Int := 0

   var account : Account := new BankAccount();

   op run == !tick() ; !trade()

   op tick == await timer < 6; timer := timer + 1; !tick()

   op trade ==
     var balance : Int;
     await timer > 5;
     timer := 0;
     account.balance(; balance);
     if balance <= 100 then account.deposit(10;) end ;
     !trade()
 end
