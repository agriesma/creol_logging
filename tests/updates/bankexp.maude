load creol-update .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op state : -> State [ctor] .
  eq state = {
    < class("BankAccount", 0) : Class | Version: 0, Inh: noInh, Param: noVid,
      Att: "bal" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( ( "bal" @ "BankAccount" ) ; int(0) ) ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "increaseBalance" : Method | Param: "sum", Att: "sum" |-> null,
              "_" |-> null, Code: assign( "bal" ; "+" ( "bal" :: "sum" ) ) ;
              return ( emp ) >,
            < "balance" : Method | Param: noVid, Att: "b" |-> null, "_" |->
              null, Code: assign( "b" ; "bal" ) ; return ( "b" ) >,
            < "deposit" : Method | Param: "sum", Att: "sum" |-> null,
              "ccglab0" |-> null, "_" |-> null,
              Code: call( "ccglab0" ; "this" ; "increaseBalance" ; "sum" ) ;
              free( "ccglab0" ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < class("Customer", 0) : Class | Version: 0, Inh: noInh, Param: noVid,
      Att: "timer" |-> null, "account" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: new( ( "account" @ "Customer" ) ; "BankAccount" ; emp ) ;
              assign( ( "timer" @ "Customer" ) ; int(0) ) ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "ccglab1" |-> null,
              "ccglab0" |-> null, "_" |-> null,
              Code: call( "ccglab0" ; "this" ; "tick" ; emp ) ;
              free( "ccglab0" ) ; call( "ccglab1" ; "this" ; "trade" ; emp )
              ; free( "ccglab1" ) ; return ( emp ) >,
            < "tick" : Method | Param: noVid, Att: "ccglab0" |-> null, "_"
              |-> null, Code: await "<" ( "timer" :: int(6) ) ;
              assign( "timer" ; "+" ( "timer" :: int(1) ) ) ;
              call( "ccglab0" ; "this" ; "tick" ; emp ) ; free( "ccglab0" ) ;
              return ( emp ) >,
            < "trade" : Method | Param: noVid, Att: "ccglab2" |-> null,
              "ccglab1" |-> null, "ccglab0" |-> null, "balance" |-> null, "_"
              |-> null, Code: await ">" ( "timer" :: int(5) ) ;
              assign( "timer" ; int(0) ) ; call( "ccglab0" ; "account" ;
              "balance" ; emp ) ; get( "ccglab0" ; "balance" ) ;
              free( "ccglab0" ) ; if "<=" ( "balance" :: int(100) ) th
              call( "ccglab1" ; "account" ; "deposit" ; int(10) ) ;
              get( "ccglab1" ; noVid ) ; free( "ccglab1" ) el skip fi ;
              call( "ccglab2" ; "this" ; "trade" ; emp ) ; free( "ccglab2" )
              ; return ( emp ) >,
      Ocnt: 0 > } .
endm

