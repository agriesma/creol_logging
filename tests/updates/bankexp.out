==========================================
frewrite in PROGRAM : main(state, "Customer", emp) .
result (sort not calculated): {
< label(ob("Customer-0"), 76) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< label(ob("Customer-0"), 79) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< label(ob("Customer-0"), 80) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< ob("BankAccount-0") : class("BankAccount",0) |
  Att: "bal" |-> int(80), "this" |-> ob("BankAccount-0"),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 10 >
 
< ob("Customer-0") : class("Customer",0) |
  Att: "account" |-> ob("BankAccount-0"), "this" |-> ob("Customer-0"), "timer"
    |-> int(4),
  Pr: {".class" |-> str("Customer"), ".label" |-> label(ob("Customer-0"), 79),
    ".method" |-> str("tick"), ".stage" |-> int(0), "_" |-> null, "caller" |->
    ob("Customer-0"), "ccglab0" |-> null | return(emp)},
  PrQ: {".class" |-> str("Customer"), ".label" |-> label(ob("Customer-0"), 80),
    ".method" |-> str("tick"), ".stage" |-> int(0), "_" |-> null, "caller" |->
    ob("Customer-0"), "ccglab0" |-> null | await "<"("timer" :: int(6)) ;
    assign("timer" ; "+"("timer" :: int(1))) ; call("ccglab0" ; "this" ; "tick"
    ; emp) ; free("ccglab0") ; return(emp)}, {".class" |-> str("Customer"),
    ".label" |-> label(ob("Customer-0"), 76), ".method" |-> str("trade"),
    ".stage" |-> int(0), "_" |-> null, "balance" |-> null, "caller" |-> ob(
    "Customer-0"), "ccglab0" |-> null, "ccglab1" |-> null, "ccglab2" |-> null |
    await ">"("timer" :: int(5)) ; assign("timer" ; int(0)) ; call("ccglab0" ;
    "account" ; "balance" ; emp) ; get("ccglab0" ; "balance") ; free("ccglab0")
    ; if "<="("balance" :: int(100)) th call("ccglab1" ; "account" ; "deposit"
    ; int(10)) ; get("ccglab1" ; noVid) ; free("ccglab1") el skip fi ; call(
    "ccglab2" ; "this" ; "trade" ; emp) ; free("ccglab2") ; return(emp)},
  Lcnt: 81 >
 
< class("BankAccount",0) : Class | Version: 0, Inh: noInh, Param: noVid, Att:
    "bal" |-> null, Mtds:
  < "balance" : Method | Param: noVid, Att: "_" |-> null, "b" |-> null, Code:
    assign("b" ; "bal") ; return("b") >, < "deposit" : Method | Param: "sum",
    Att: "_" |-> null, "ccglab0" |-> null, "sum" |-> null, Code: call("ccglab0"
    ; "this" ; "increaseBalance" ; "sum") ; free("ccglab0") ; return(emp) >, <
    "increaseBalance" : Method | Param: "sum", Att: "_" |-> null, "sum" |->
    null, Code: assign("bal" ; "+"("bal" :: "sum")) ; return(emp) >, < "init" :
    Method | Param: noVid, Att: "_" |-> null, Code: assign("bal" @
    "BankAccount" ; int(0)) ; return(emp) >, < "run" : Method | Param: noVid,
    Att: "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 1 >
 
< class("Customer",0) : Class | Version: 0, Inh: noInh, Param: noVid, Att:
    "account" |-> null, "timer" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: new("account" @
    "Customer" ; "BankAccount" ; emp) ; assign("timer" @ "Customer" ; int(0)) ;
    return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null,
    "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0" ; "this" ;
    "tick" ; emp) ; free("ccglab0") ; call("ccglab1" ; "this" ; "trade" ; emp)
    ; free("ccglab1") ; return(emp) >, < "tick" : Method | Param: noVid, Att:
    "_" |-> null, "ccglab0" |-> null, Code: await "<"("timer" :: int(6)) ;
    assign("timer" ; "+"("timer" :: int(1))) ; call("ccglab0" ; "this" ; "tick"
    ; emp) ; free("ccglab0") ; return(emp) >, < "trade" : Method | Param:
    noVid, Att: "_" |-> null, "balance" |-> null, "ccglab0" |-> null, "ccglab1"
    |-> null, "ccglab2" |-> null, Code: await ">"("timer" :: int(5)) ; assign(
    "timer" ; int(0)) ; call("ccglab0" ; "account" ; "balance" ; emp) ; get(
    "ccglab0" ; "balance") ; free("ccglab0") ; if "<="("balance" :: int(100))
    th call("ccglab1" ; "account" ; "deposit" ; int(10)) ; get("ccglab1" ;
    noVid) ; free("ccglab1") el skip fi ; call("ccglab2" ; "this" ; "trade" ;
    emp) ; free("ccglab2") ; return(emp) >,
  Ocnt: 1 >
}
Bye.
